protocol/scripts/iaviewer/iaviewer.go

```diff
@@ -8,59 +8,177 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
-	"strconv"
+	"regexp"
 	"strings"
 
 	dbm "github.com/cometbft/cometbft-db"
-
 	"github.com/cosmos/iavl"
+	"github.com/spf13/cobra"
 )
 
 // TODO: make this configurable?
 const (
 	DefaultCacheSize int = 10000
 )
 
+var (
+	prefixRe *regexp.Regexp = regexp.MustCompile("(s/k:[[:alpha:]]+/).*")
+)
+
+var rootCmd = &cobra.Command{
+	Use:          "iaviewer",
+	SilenceUsage: true,
+}
+
 func main() {
-	args := os.Args[1:]
-	if len(args) < 3 || (args[0] != "data" && args[0] != "shape" && args[0] != "versions") {
-		fmt.Fprintln(os.Stderr, "Usage: iaviewer <data|shape|versions> <leveldb dir> <prefix> [version number]")
-		fmt.Fprintln(os.Stderr, "<prefix> is the prefix of db, and the iavl tree of different modules in cosmos-sdk uses ")
-		fmt.Fprintln(os.Stderr, "different <prefix> to identify, just like \"s/k:gov/\" represents the prefix of gov module")
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "iaviewer got error: '%s'", err)
 		os.Exit(1)
 	}
+}
 
-	version := 0
-	if len(args) == 4 {
-		var err error
-		version, err = strconv.Atoi(args[3])
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Invalid version number: %s\n", err)
-			os.Exit(1)
-		}
+func init() {
+	rootCmd.AddCommand(DataAllCmd())
+	rootCmd.AddCommand(DataCmd())
+	rootCmd.AddCommand(ShapeCmd())
+	rootCmd.AddCommand(VersionsCmd())
+}
+
+func AddPathFlag(cmd *cobra.Command) {
+	cmd.Flags().String("path", "", "path to leveldb dir")
+	if err := cmd.MarkFlagRequired("path"); err != nil {
+		panic(err)
 	}
+}
 
-	tree, err := ReadTree(args[1], version, []byte(args[2]))
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error reading data: %s\n", err)
-		os.Exit(1)
+func AddPrefixFlag(cmd *cobra.Command) {
+	cmd.Flags().String("prefix", "", "prefix of the db e.g. \"s/k:gov/\"")
+	if err := cmd.MarkFlagRequired("prefix"); err != nil {
+		panic(err)
 	}
+}
 
-	switch args[0] {
-	case "data":
-		PrintKeys(tree)
-		hash, err := tree.Hash()
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Error hashing tree: %s\n", err)
-			os.Exit(1)
-		}
-		fmt.Printf("Hash: %X\n", hash)
-		fmt.Printf("Size: %X\n", tree.Size())
-	case "shape":
-		PrintShape(tree)
-	case "versions":
-		PrintVersions(tree)
+func AddVersionFlag(cmd *cobra.Command) {
+	cmd.Flags().Uint64("version", 0, "version of the tree (default: latest version)")
+}
+
+func DataAllCmd() *cobra.Command {
```

Comment:
for my understanding - how did this work before `prefix` option was added? kind of surprised that this is not already supported

Reply:
It doesn't work. If you don't pass prefix in, it will assume that all the leveldb keys are unprefixed and try to interpret it as an iavl tree. But for dbs created by cosmos, all the keys are prefixed and not ready to be interpreted as iavl tree unless the prefix is removed. So you'd just find no values unless you pass in a prefix of a tree.

Reply:
🤯